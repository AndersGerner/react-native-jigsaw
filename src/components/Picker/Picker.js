import * as React from "react";
import { withTheme } from "../../core/theming";
import PickerComponent from "./PickerComponent";

import {
  COMPONENT_TYPES,
  FORM_TYPES,
  FIELD_NAME,
} from "../../core/component-types";

class Picker extends React.Component {
  static defaultProps = {
    options: [],
    type: "underline",
    disabled: false,
    error: false,
  };

  state = {
    pickerVisible: false,
  };

  onValueChange = (itemValue, itemIndex) => {
    const { placeholder, onChange } = this.props;

    if (placeholder && itemIndex === 0) {
      return;
    }

    onChange && onChange(itemValue, itemIndex);
  };

  render() {
    const { placeholder, options, value, ...props } = this.props;

    const pickerOptions = placeholder
      ? [{ value: placeholder, label: placeholder }, ...options]
      : options;

    return (
      <PickerComponent
        {...props}
        selectedValue={value}
        placeholder={placeholder}
        options={pickerOptions}
        onValueChange={this.onValueChange}
      />
    );
  }
}

export default withTheme(Picker);

const SEED_DATA_PROPS = {
  label: {
    label: "Label",
    description: "The label to be displayed on the picker",
    formType: FORM_TYPES.string,
    defaultValue: "Make",
    editable: true,
    required: true,
  },
  placeholder: {
    label: "Placeholder",
    description: "The placeholder text of the picker",
    formType: FORM_TYPES.string,
    defaultValue: "Select a make...",
    editable: true,
    required: false,
  },
  assistiveText: {
    label: "Assistive text",
    description: "Helper text to display below the picker",
    formType: FORM_TYPES.string,
    defaultValue: null,
    editable: true,
    required: false,
  },
  options: {
    label: "Options",
    description:
      "Array of picker options. An array of objects containing a label and value.",
    editable: true,
    formType: FORM_TYPES.array,
    defaultValue: null,
  },
  disabled: {
    label: "Disabled",
    description:
      "Whether the picker should be disabled. Will prevent selection and show a greyed out state.",
    formType: FORM_TYPES.boolean,
    defaultValue: false,
    editable: true,
  },
  error: {
    label: "Error",
    description: "Whether the picker should display the error state",
    formType: FORM_TYPES.boolean,
    defaultValue: false,
    editable: true,
  },
  leftIconName: {
    label: "Left icon name",
    description: "The icon to display on the left",
    formType: FORM_TYPES.icon,
    defaultValue: null,
    editable: true,
  },
  leftIconMode: {
    label: "Left icon mode",
    description:
      "The mode of the icon to display on the left. 'inset' or 'outset'.",
    formType: FORM_TYPES.flatArray,
    defaultValue: "inset",
    options: ["inset", "outset"],
    editable: true,
    required: true,
  },
  rightIconName: {
    label: "Right icon name",
    description: "The icon to display on the right",
    formType: FORM_TYPES.icon,
    defaultValue: null,
    editable: true,
  },
  fieldName: {
    ...FIELD_NAME,
    defaultValue: "pickerValue",
  },
};

export const SEED_DATA = [
  {
    name: "Picker - Solid",
    tag: "Picker",
    description: "A picker with a solid border",
    category: COMPONENT_TYPES.field,
    preview_image_url: "{CLOUDINARY_URL}/Picker.png",
    supports_list_render: false,
    props: {
      ...SEED_DATA_PROPS,
      type: {
        formType: FORM_TYPES.string,
        defaultValue: "solid",
        editable: false,
      },
    },
    layout: {},
  },
  {
    name: "Picker - Underline",
    tag: "Picker",
    description: "A picker with an underline",
    category: COMPONENT_TYPES.field,
    preview_image_url: "{CLOUDINARY_URL}/Picker.png",
    supports_list_render: false,
    props: {
      ...SEED_DATA_PROPS,
      type: {
        formType: FORM_TYPES.string,
        defaultValue: "underline",
        editable: false,
      },
    },
    layout: {},
  },
];
