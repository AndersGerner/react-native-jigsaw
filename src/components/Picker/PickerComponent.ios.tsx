import * as React from "react";
import {
  View,
  SafeAreaView,
  StyleSheet,
  Picker as NativePicker,
  StyleProp,
  ViewStyle,
} from "react-native";
import { withTheme } from "../../core/theming";

import Portal from "../Portal/Portal";
import Button from "../Button";
import TextField, { Props as TextFieldProps } from "../TextField";
import Touchable from "../Touchable";

interface PickerOption {
  value: string;
  label: string;
}

interface Props extends TextFieldProps {
  style?: StyleProp<ViewStyle>;
  options: PickerOption[];
  placeholder?: string;
  selectedValue: string;
  disabled?: boolean;
  onValueChanged: (value: string, index: number) => void;
}

const Picker: React.FC<Props> = ({
  style,
  options,
  placeholder,
  selectedValue,
  disabled = false,
  onValueChanged = () => {},
  theme: { colors },
  ...props
}) => {
  const textField = React.useRef<typeof TextField | undefined>(undefined);
  const [pickerVisible, setIsPickerVisible] = React.useState(false);

  const toggleVisibility = () => {
    setIsPickerVisible(!pickerVisible);
    textField.current.toggleFocus();
  };

  return (
    <View style={[styles.container, style]}>
      <Touchable disabled={disabled} onPress={toggleVisibility}>
        <TextField
          {...props}
          value={selectedValue}
          placeholder={placeholder}
          ref={textField}
          disabled={disabled}
          pointerEvents="none"
        />
      </Touchable>
      {pickerVisible && (
        <Portal>
          <View style={[styles.picker, { backgroundColor: colors.divider }]}>
            <SafeAreaView style={styles.pickerContainer}>
              <Button
                type="text"
                onPress={toggleVisibility}
                style={styles.closeButton}
              >
                Close
              </Button>
              <NativePicker
                selectedValue={selectedValue}
                onValueChange={onValueChanged}
              >
                {options.map((o) => (
                  <NativePicker.Item
                    label={o.label}
                    value={o.value}
                    key={o.value}
                  />
                ))}
              </NativePicker>
            </SafeAreaView>
          </View>
        </Portal>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignSelf: "stretch",
  },
  picker: {
    position: "absolute",
    bottom: 0,
    left: 0,
    right: 0,
    flexDirection: "row",
    justifyContent: "center",
  },
  pickerContainer: { flexDirection: "column", width: "100%" },
  closeButton: {
    alignSelf: "flex-end",
  },
});

export default withTheme(Picker);
