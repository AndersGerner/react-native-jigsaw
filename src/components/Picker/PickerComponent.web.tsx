import * as React from "react";
import {
  View,
  Picker as NativePicker,
  StyleSheet,
  StyleProp,
  ViewStyle,
} from "react-native";
import { withTheme } from "../../core/theming";

import TextField, { Props as TextFieldProps } from "../TextField";
import Touchable from "../Touchable";

interface PickerOption {
  value: string;
  label: string;
}

interface Props extends TextFieldProps {
  style?: StyleProp<ViewStyle>;
  options: PickerOption[];
  placeholder?: string;
  selectedValue: string;
  disabled?: boolean;
  onValueChanged: (value: string, index: number) => void;
}

const Picker: React.FC<Props> = ({
  style,
  options,
  placeholder,
  selectedValue,
  disabled = false,
  onValueChanged: onValueChangeOverride = () => {},
  ...props
}) => {
  const textField = React.useRef<typeof TextField | undefined>(undefined);

  const onValueChange = (itemValue: string, itemIndex: number) => {
    toggleFocus();
    onValueChangeOverride(itemValue, itemIndex);
  };

  const toggleFocus = () => {
    if (!disabled) {
      textField.current.toggleFocus();
    }
  };

  return (
    <Touchable
      disabled={disabled}
      onPress={toggleFocus}
      style={[styles.container, style]}
    >
      <View>
        <NativePicker
          enabled={!disabled}
          selectedValue={selectedValue}
          onValueChange={onValueChange}
          style={{
            flex: 1,
            opacity: 0,
            position: "absolute",
            top: 0,
            left: 0,
            right: 0,
            bottom: 0,
          }}
        >
          {options.map((o) => (
            <NativePicker.Item label={o.label} value={o.value} key={o.value} />
          ))}
        </NativePicker>
        <View pointerEvents="none">
          <TextField
            {...props}
            value={selectedValue}
            placeholder={placeholder}
            ref={textField}
            disabled={disabled}
          />
        </View>
      </View>
    </Touchable>
  );
};

const styles = StyleSheet.create({
  container: {
    alignSelf: "stretch",
  },
});

export default withTheme(Picker);
