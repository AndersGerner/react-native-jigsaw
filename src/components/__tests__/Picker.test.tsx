import * as React from "react";
import renderer from "react-test-renderer";
import Picker from "../Picker/Picker";

it("renders picker", () => {
  const tree = renderer
    .create(
      <Picker
        value="value1"
        options={[{ label: "label1", value: "value1" }]}
        onValueChange={() => {}}
        selectedValue="value1"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
