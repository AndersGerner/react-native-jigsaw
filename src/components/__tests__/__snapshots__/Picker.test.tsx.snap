// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders picker 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
      },
      undefined,
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignSelf": "stretch",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "borderTopLeftRadius": 6,
              "borderTopRightRadius": 6,
              "marginTop": 16,
              "paddingBottom": 12,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgba(165, 173, 183, 1)",
                "bottom": 0,
                "height": 2,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": Array [
                  Object {
                    "scaleY": 0.5,
                  },
                ],
              }
            }
          />
          <TextInput
            allowFontScaling={true}
            editable={true}
            multiline={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholderTextColor="rgba(165, 173, 183, 1)"
            pointerEvents="none"
            rejectResponderTermination={true}
            selectionColor="rgba(90, 69, 255, 1)"
            style={
              Array [
                Object {
                  "flexGrow": 1,
                  "justifyContent": "center",
                  "margin": 0,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                Object {
                  "backgroundColor": "transparent",
                  "color": "rgba(18, 20, 44, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "height": 26,
                  "letterSpacing": 0,
                  "paddingLeft": 0,
                  "paddingRight": 12,
                  "paddingVertical": 0,
                },
                Object {},
              ]
            }
            underlineColorAndroid="transparent"
            value="value1"
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
